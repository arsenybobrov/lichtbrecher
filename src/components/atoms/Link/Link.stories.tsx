import React from 'react';
import Link from './Link';

export default {
  title: 'Atoms/Link',
  component: Link,
};

const externalLink = {
  link_type: 'Web',
  url: 'https://example.com',
};

const internalLink = {
  id: 'X8kLmBIAACkAn_Za',
  type: 'page',
  tags: [],
  slug: 'irgendeine-themenseite',
  lang: 'de-de',
  uid: 'irgendeine-themenseite',
  link_type: 'Document',
  isBroken: false,
};

export const Primary = () => (
  <Link url={externalLink} text="external link" title="click me" skin="primary" />
);

export const PrimaryInternalLink = () => (
  <Link
    url={internalLink}
    text="internal link"
    title="click me"
    skin="primary"
  />
);

export const PrimaryAnchorLink = () => (
  <Link text="anchor link" title="click me" skin="primary" anchor_id="funky_lement" />
);

export const PrimaryBig = () => (
  <Link url={externalLink} text="external link" title="click me" skin="primaryBig" />
);

// TODO: make work in storybook
export const PrimaryBigActiveLink = () => (
  <Link
    url={internalLink}
    text="internal link"
    title="click me"
    skin="primaryBig"
    showActive
  />
);

export const Copytext = () => (
  <Link url={externalLink} text="external link" title="click me" skin="copytext" />
);

export const Tertiary = () => (
  <Link url={externalLink} text="external link" title="click me" skin="tertiary" />
);

export const Primary700 = () => (
  <Link url={externalLink} text="external link" title="click me" skin="primary700" />
);
