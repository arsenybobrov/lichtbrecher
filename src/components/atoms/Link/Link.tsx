import React from 'react';
import Router from 'next/router';
import get from 'lodash/get';
import linkResolver from '../../../../prismic/helper/linkResolver';
import { PrismicLink } from '../../../../prismic/types';

interface LinkProps {
  text: string,
  url: PrismicLink,
  title: string,
}

const Link: React.FC<LinkProps> = ({
  text, url, title = text,
}) => {
  const handleClick = (
    e: React.MouseEvent<HTMLAnchorElement>,
    href: string,
    as: string,
    options?: any
  ) => {
    e.preventDefault();
    setTimeout(() => {
      Router.push(href, as, options);
    }, 0);
  };

  if (url) {
    if (url?.link_type !== 'Document') {
      return (
        <a
          href={linkResolver(url)}
          title={title}
          target={get(url, 'target', null)}
          rel={get(url, 'target', '_self') === '_blank' ? 'noopener noreferrer' : undefined}
        >
          {text}
        </a>
      );
    }

    return (
      <a
        href={linkResolver(url)}
        title={title}
        onClick={(e) => handleClick(e, '/', linkResolver(url))}
      >
        {text}
      </a>
    );
  }
  return <></>;
};

export default Link;
