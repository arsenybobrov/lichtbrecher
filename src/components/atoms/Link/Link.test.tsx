import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Link from './Link';
import {
  prismicExternalLinkMock,
  prismicInternalHomepageLinkMock,
  prismicInternalPageLinkMock,
} from '../../../../prismic/helper/tests/linkResolver.test';

export const NextLinkInternalPageLinkMock = {
  url: prismicInternalPageLinkMock,
  title: 'here comes the title',
  text: 'Click me',
};

export const NextLinkInternalPageLinkMockWithChildren = {
  url: prismicInternalPageLinkMock,
  title: 'here comes the title',
};

export const NextLinkInternalHomepageLinkMock = {
  url: prismicInternalHomepageLinkMock,
  title: 'here comes the title',
  text: 'Click me',
};

export const NextLinkExternalLinkMock = {
  url: prismicExternalLinkMock,
  text: 'Click me',
};

export const NextLinkExternalLinkMockBlank = {
  url: prismicExternalLinkMock,
  text: 'Click me',
  target: '_blank',
};

describe('nextLink snapshot', () => {
  it.each`
    name                                               | mock
    ${'internal Link to a page'}                       | ${NextLinkInternalPageLinkMock}
    ${'internal Link to the homepage'}                 | ${NextLinkInternalHomepageLinkMock}
    ${'external link'}                                 | ${NextLinkExternalLinkMock}
    ${'external link with target _blank'}              | ${NextLinkExternalLinkMockBlank}
  `('$name', ({ mock }) => {
  const tree = renderer.create(<Link {...mock} />).toJSON();
  expect(tree).toMatchSnapshot();
});
  it.each`
    name                                       | mock
    ${'internal Link to a page with children'} | ${NextLinkInternalPageLinkMockWithChildren}
  `('$name', ({ mock }) => {
  const tree = renderer.create(<Link {...mock}>Click me!</Link>).toJSON();
  expect(tree).toMatchSnapshot();
});
});
