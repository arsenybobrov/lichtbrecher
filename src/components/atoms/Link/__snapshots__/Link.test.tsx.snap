// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nextLink snapshot external link 1`] = `
<a
  href="https://example.com"
  rel="noopener noreferrer"
  target="_blank"
  title="Click me"
>
  Click me
</a>
`;

exports[`nextLink snapshot external link with target _blank 1`] = `
<a
  href="https://example.com"
  rel="noopener noreferrer"
  target="_blank"
  title="Click me"
>
  Click me
</a>
`;

exports[`nextLink snapshot internal Link to a page 1`] = `
<a
  href="/page-two"
  onClick={[Function]}
  title="here comes the title"
>
  Click me
</a>
`;

exports[`nextLink snapshot internal Link to a page with children 1`] = `
<a
  href="/page-two"
  onClick={[Function]}
  title="here comes the title"
>
  Click me!
</a>
`;

exports[`nextLink snapshot internal Link to the homepage 1`] = `
<a
  href="/"
  onClick={[Function]}
  title="here comes the title"
>
  Click me
</a>
`;
